library(dplyr)

# calculate tpm values for bowtie results

calculate_TPM <- function(counts_file, output_file = "tpm_values.txt") {
  # Step 1: Load the raw counts file generated by featureCounts
  counts <- read.table(counts_file, header = TRUE, sep = "\t")
  colnames(counts)[7] <- "counts"
  # Step 2: Calculate RPK (Reads Per Kilobase)
  # Convert gene length to kilobases (kb)
  counts$Length_kb <- counts$Length / 1000 # Convert length to kb
  counts$RPK <- counts$counts / counts$Length_kb
  # Step 3: Calculate CPM (Counts Per Million)
  # First, calculate the total reads in the sample
  total_reads <- sum(counts$counts)
  # Now calculate CPM for each gene
  counts$CPM <- (counts$counts / total_reads) * 1e6
  # Step 4: Calculate TPM (Transcripts Per Million)
  # Calculate TPM by normalizing CPM by gene length
  counts$TPM <- (counts$CPM / counts$Length_kb)
  # Print the first few rows with TPM values
  head(counts)
  # Step 5: Save the result in a new file (optional)
  write.table(counts,
    file = output_file, sep = "\t", row.names = FALSE,
    quote = FALSE
  )
  return(counts) # Return the dataframe with TPM values
}

calculate_TPM("../results/control/9929263/counts.txt", "../results/control/9929263/tpm_values.txt")
calculate_TPM("../results/control/9929264/counts.txt", "../results/control/9929264/tpm_values.txt")
calculate_TPM("../results/control/9929273/counts.txt", "../results/control/9929273/tpm_values.txt")
calculate_TPM("../results/condition1/9929265/counts.txt", "../results/condition1/9929265/tpm_values.txt")
calculate_TPM("../results/condition1/9929271/counts.txt", "../results/condition1/9929271/tpm_values.txt")
calculate_TPM("../results/condition1/9929274/counts.txt", "../results/condition1/9929274/tpm_values.txt")
calculate_TPM("../results/condition2/9929272/counts.txt", "../results/condition2/9929272/tpm_values.txt")
calculate_TPM("../results/condition2/9929279/counts.txt", "../results/condition2/9929279/tpm_values.txt")
calculate_TPM("../results/condition2/9929281/counts.txt", "../results/condition2/9929281/tpm_values.txt")

# calculate tpm values for kallisto results

ctl_9929263 <- read.delim("../results/control/9929263/abundance.tsv")
ctl_9929264 <- read.delim("../results/control/9929264/abundance.tsv")
ctl_9929273 <- read.delim("../results/control/9929273/abundance.tsv")
cn1_9929265 <- read.delim("../results/condition1/9929265/abundance.tsv")
cn1_9929271 <- read.delim("../results/condition1/9929271/abundance.tsv")
cn1_9929274 <- read.delim("../results/condition1/9929274/abundance.tsv")
cn2_9929272 <- read.delim("../results/condition2/9929272/abundance.tsv")
cn2_9929279 <- read.delim("../results/condition2/9929279/abundance.tsv")
cn2_9929281 <- read.delim("../results/condition2/9929281/abundance.tsv")

reduce_name <- function(id) {
  sub("[_].*$", "", id)
}

ctl_9929263["gene"] <- apply(ctl_9929263["target_id"], 2, reduce_name)
ctl_9929264["gene"] <- apply(ctl_9929264["target_id"], 2, reduce_name)
ctl_9929273["gene"] <- apply(ctl_9929273["target_id"], 2, reduce_name)
cn1_9929265["gene"] <- apply(cn1_9929265["target_id"], 2, reduce_name)
cn1_9929271["gene"] <- apply(cn1_9929271["target_id"], 2, reduce_name)
cn1_9929274["gene"] <- apply(cn1_9929274["target_id"], 2, reduce_name)
cn2_9929272["gene"] <- apply(cn2_9929272["target_id"], 2, reduce_name)
cn2_9929279["gene"] <- apply(cn2_9929279["target_id"], 2, reduce_name)
cn2_9929281["gene"] <- apply(cn2_9929281["target_id"], 2, reduce_name)

ctl_9929263_agg <- aggregate(tpm ~ gene, data = ctl_9929263, sum)
ctl_9929264_agg <- aggregate(tpm ~ gene, data = ctl_9929264, sum)
ctl_9929273_agg <- aggregate(tpm ~ gene, data = ctl_9929273, sum)
cn1_9929265_agg <- aggregate(tpm ~ gene, data = cn1_9929265, sum)
cn1_9929271_agg <- aggregate(tpm ~ gene, data = cn1_9929271, sum)
cn1_9929274_agg <- aggregate(tpm ~ gene, data = cn1_9929274, sum)
cn2_9929272_agg <- aggregate(tpm ~ gene, data = cn2_9929272, sum)
cn2_9929279_agg <- aggregate(tpm ~ gene, data = cn2_9929279, sum)
cn2_9929281_agg <- aggregate(tpm ~ gene, data = cn2_9929281, sum)

write.table(ctl_9929263_agg, "../results/control/9929263/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(ctl_9929264_agg, "../results/control/9929264/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(ctl_9929273_agg, "../results/control/9929273/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn1_9929265_agg, "../results/condition1/9929265/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn1_9929271_agg, "../results/condition1/9929271/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn1_9929274_agg, "../results/condition1/9929274/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn2_9929272_agg, "../results/condition2/9929272/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn2_9929279_agg, "../results/condition2/9929279/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
write.table(cn2_9929281_agg, "../results/condition2/9929281/tpm_values_kallisto.txt", sep = "\t", row.names = FALSE, quote = FALSE)
require(ggplot2)
require(dplyr)
require(reshape2)

# creating grouped boxplots

kallisto_data <- list(
  ctl_9929263_k = read.delim("../results/control/9929263/tpm_values_kallisto.txt"),
  ctl_9929264_k = read.delim("../results/control/9929264/tpm_values_kallisto.txt"),
  ctl_9929273_k = read.delim("../results/control/9929273/tpm_values_kallisto.txt"),
  cn1_9929265_k = read.delim("../results/condition1/9929265/tpm_values_kallisto.txt"),
  cn1_9929271_k = read.delim("../results/condition1/9929271/tpm_values_kallisto.txt"),
  cn1_9929274_k = read.delim("../results/condition1/9929274/tpm_values_kallisto.txt"),
  cn2_9929272_k = read.delim("../results/condition2/9929272/tpm_values_kallisto.txt"),
  cn2_9929279_k = read.delim("../results/condition2/9929279/tpm_values_kallisto.txt"),
  cn2_9929281_k = read.delim("../results/condition2/9929281/tpm_values_kallisto.txt")
)

for (i in seq_along(kallisto_data)) {
  kallisto_data[[i]]$group <- rep("kallisto", length(kallisto_data[[i]]$tpm))
  kallisto_data[[i]]$seq <- substring(names(kallisto_data)[i], 5, 11)
  kallisto_data[[i]]$exp <- substring(names(kallisto_data)[i], 1, 3)
}

bowtie2_data <- list(
  ctl_9929263_b = read.delim("../results/control/9929263/tpm_values.txt"),
  ctl_9929264_b = read.delim("../results/control/9929264/tpm_values.txt"),
  ctl_9929273_b = read.delim("../results/control/9929273/tpm_values.txt"),
  cn1_9929265_b = read.delim("../results/condition1/9929265/tpm_values.txt"),
  cn1_9929271_b = read.delim("../results/condition1/9929271/tpm_values.txt"),
  cn1_9929274_b = read.delim("../results/condition1/9929274/tpm_values.txt"),
  cn2_9929272_b = read.delim("../results/condition2/9929272/tpm_values.txt"),
  cn2_9929279_b = read.delim("../results/condition2/9929279/tpm_values.txt"),
  cn2_9929281_b = read.delim("../results/condition2/9929281/tpm_values.txt")
)

for (i in seq_along(bowtie2_data)) {
  names(bowtie2_data[[i]])[11] <- "tpm"
  names(bowtie2_data[[i]])[1] <- "gene"
  bowtie2_data[[i]]$group <- rep("bowtie2", length(bowtie2_data[[i]]$tpm))
  bowtie2_data[[i]]$seq <- substring(names(bowtie2_data)[i], 5, 11)
  bowtie2_data[[i]]$exp <- substring(names(bowtie2_data)[i], 1, 3)
}

all_data <- rbind(bind_rows(bowtie2_data)[, c("gene", "tpm", "group", "seq", "exp")], bind_rows(kallisto_data))

ggplot(data = all_data, aes(x = interaction(group, exp), y = tpm, fill = group)) +
  geom_boxplot() +
  labs(x = "Gene expression program and experiment", y = "TPM values") +
  theme_minimal() +
  scale_fill_brewer(palette = "Pastel1")

all_data$tpm_scaled <- scale(all_data$tpm)

all_data$tpm_log <- log1p(all_data$tpm)

ggplot(data = all_data, aes(x = interaction(group, exp), y = tpm_log, fill = group)) +
  geom_boxplot() +
  labs(x = "Gene expression program and experiment", y = "TPM values") +
  theme_minimal() +
  scale_fill_brewer(palette = "Pastel1")

ggplot(data = all_data, aes(x = interaction(group, exp), y = tpm_log, fill = group)) +
  geom_violin() +
  labs(x = "Gene expression program and experiment", y = "TPM values") +
  theme_minimal() +
  scale_fill_brewer(palette = "Pastel1")

# computing log fold changes

all_data$lfc1 <- NA
all_data$lfc2 <- NA
all_data$lfc3 <- NA # change from cn2 to cn1

all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "lfc1"] <-
  log2(all_data[all_data$group == "kallisto" & all_data$exp == "cn1", "tpm"] /
    all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "tpm"])
all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "lfc2"] <-
  log2(all_data[all_data$group == "kallisto" & all_data$exp == "cn2", "tpm"] /
    all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "tpm"])
all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "lfc3"] <-
  log2(all_data[all_data$group == "kallisto" & all_data$exp == "cn2", "tpm"] /
    all_data[all_data$group == "kallisto" & all_data$exp == "cn1", "tpm"])

all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "lfc1"] <-
  log2(all_data[all_data$group == "bowtie2" & all_data$exp == "cn1", "tpm"] /
    all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "tpm"])
all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "lfc2"] <-
  log2(all_data[all_data$group == "bowtie2" & all_data$exp == "cn2", "tpm"] /
    all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "tpm"])
all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "lfc3"] <-
  log2(all_data[all_data$group == "bowtie2" & all_data$exp == "cn2", "tpm"] /
    all_data[all_data$group == "bowtie2" & all_data$exp == "cn1", "tpm"])

plot_lfc <- function(data, lfc) {
  plot <- ggplot(data, aes(x = data[[lfc]], fill = group)) +
    geom_density(alpha = 0.5) +
    labs(x = lfc) +
    xlim(-8, 8) +
    theme_minimal()
  ggsave(paste("../results/", lfc, ".pdf"), plot = plot, device = "pdf")
}

plot_lfc(all_data, "lfc1")
plot_lfc(all_data, "lfc2")
plot_lfc(all_data, "lfc3")

head(all_data[abs(all_data$lfc1) > 2 & !is.na(all_data$lfc1), ]$gene)

# comparison kallisto and bowtie2

intersect_genes_ctl <- intersect(
  all_data[all_data$group == "kallisto" & all_data$exp == "ctl", "gene"],
  all_data[all_data$group == "bowtie2" & all_data$exp == "ctl", "gene"]
)
intersect_genes_cn1 <- intersect(
  all_data[all_data$group == "kallisto" & all_data$exp == "cn1", "gene"],
  all_data[all_data$group == "bowtie2" & all_data$exp == "cn1", "gene"]
)
intersect_genes_cn2 <- intersect(
  all_data[all_data$group == "kallisto" & all_data$exp == "cn2", "gene"],
  all_data[all_data$group == "bowtie2" & all_data$exp == "cn2", "gene"]
)

all_data$diff <- NA

for (gene in intersect_genes_ctl) {
  all_data[all_data$gene == gene & all_data$exp == "ctl", "diff"] <- (
    all_data[all_data$gene == gene & all_data$exp == "ctl" & all_data$group == "kallisto", "tpm"] -
      all_data[all_data$gene == gene & all_data$exp == "ctl" & all_data$group == "bowtie2", "tpm"]
  )
}

for (gene in intersect_genes_cn1) {
  all_data[all_data$gene == gene & all_data$exp == "cn1", "diff"] <- (
    all_data[all_data$gene == gene & all_data$exp == "cn1" & all_data$group == "kallisto", "tpm"] -
      all_data[all_data$gene == gene & all_data$exp == "cn1" & all_data$group == "bowtie2", "tpm"]
  )
}

for (gene in intersect_genes_cn2) {
  all_data[all_data$gene == gene & all_data$exp == "cn2", "diff"] <- (
    all_data[all_data$gene == gene & all_data$exp == "cn2" & all_data$group == "kallisto", "tpm"] -
      all_data[all_data$gene == gene & all_data$exp == "cn2" & all_data$group == "bowtie2", "tpm"]
  )
}

ggplot(all_data, aes(x = diff, fill = exp)) +
  geom_density(alpha = 0.5) +
  xlim(-250, 250) +
  theme_minimal()

all_data$frac <- NA

for (gene in intersect_genes_ctl) {
  all_data[all_data$gene == gene & all_data$exp == "ctl", "frac"] <- (
    log2(all_data[all_data$gene == gene & all_data$exp == "ctl" & all_data$group == "kallisto", "tpm"] /
      all_data[all_data$gene == gene & all_data$exp == "ctl" & all_data$group == "bowtie2", "tpm"])
  )
}

for (gene in intersect_genes_cn1) {
  all_data[all_data$gene == gene & all_data$exp == "cn1", "frac"] <- (
    log2(all_data[all_data$gene == gene & all_data$exp == "cn1" & all_data$group == "kallisto", "tpm"] /
      all_data[all_data$gene == gene & all_data$exp == "cn1" & all_data$group == "bowtie2", "tpm"])
  )
}

for (gene in intersect_genes_cn2) {
  log2(all_data[all_data$gene == gene & all_data$exp == "cn2", "frac"] <- (
    all_data[all_data$gene == gene & all_data$exp == "cn2" & all_data$group == "kallisto", "tpm"] /
      all_data[all_data$gene == gene & all_data$exp == "cn2" & all_data$group == "bowtie2", "tpm"]))
}

ggplot(all_data, aes(x = frac, fill = exp)) +
  geom_density(alpha = 0.5) +
  xlim(-10, 10) +
  theme_minimal()



jaccard_lfc1 <- length(intersect(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc1) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc1) >= 2, "gene"]
)) / length(union(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc1) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc1) >= 2, "gene"]
))

jaccard_lfc2 <- length(intersect(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc2) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc2) >= 2, "gene"]
)) / length(union(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc2) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc2) >= 2, "gene"]
))

jaccard_lfc3 <- length(intersect(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc3) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc3) >= 2, "gene"]
)) / length(union(
  all_data[all_data$group == "kallisto" & abs(all_data$lfc3) >= 2, "gene"],
  all_data[all_data$group == "bowtie2" & abs(all_data$lfc3) >= 2, "gene"]
))

corr_df <- data.frame(
  ctl_9929263 = all_data[all_data$group == "kallisto" & all_data$seq == "9929263", "tpm"],
  ctl_9929264 = all_data[all_data$group == "kallisto" & all_data$seq == "9929264", "tpm"],
  ctl_9929273 = all_data[all_data$group == "kallisto" & all_data$seq == "9929273", "tpm"],
  cn_1_9929265 = all_data[all_data$group == "kallisto" & all_data$seq == "9929265", "tpm"],
  cn_1_9929271 = all_data[all_data$group == "kallisto" & all_data$seq == "9929271", "tpm"],
  cn_1_9929274 = all_data[all_data$group == "kallisto" & all_data$seq == "9929274", "tpm"],
  cn_2_9929272 = all_data[all_data$group == "kallisto" & all_data$seq == "9929272", "tpm"],
  cn_2_9929279 = all_data[all_data$group == "kallisto" & all_data$seq == "9929279", "tpm"],
  cn_2_9929281 = all_data[all_data$group == "kallisto" & all_data$seq == "9929281", "tpm"]
)

corr_df <- melt(cor(corr_df))

ggplot(corr_df, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
